src/backend/distributed/transaction/remote_transaction.c

```diff
@@ -444,8 +444,8 @@ FinishRemoteTransactionAbort(MultiConnection *connection)
 
 	PQclear(result);
 
-	result = GetRemoteCommandResult(connection, dontRaiseErrors);
-	Assert(!result);
+	/* the transaction is already aborted, clear any pending activity if exists */
+	ForgetResults(connection);
```

Comment:
This entire function should perhaps just be:

```c
if (!ClearResults(connection, false))
{
    ShutdownConnection(connection);
}
```

The code above tries to give some meaningful warning if the `ABORT` fails, but it might not actually be checking the result of the abort, but rather the result of the preceding command. In the 1PC case the an abort failure has 0 consequences, but we should probably close the connection because it is in a weird state. In the 2PC case we might need to wait until transaction recovery before the locks are released. 

We could keep the `WarnAboutLeakedPreparedTransaction` since we cannot be in this situation where we send `ROLLBACK PREPARED` while another command is still in progress.

maybe something like:

```c
bool raiseErrors = false;

if (transaction->transactionState == REMOTE_TRANS_2PC_ABORTING)
{
    PGresult *result = GetRemoteCommandResult(connection, raiseErrors);
    if (!IsResponseOK(result))
    {
        bool isCommit = false;
        WarnAboutLeakedPreparedTransaction(connection, isCommit);
    }

    PQclear(result);
}

if (!ClearResults(connection, raiseErrors))
{
    ShutdownConnection(connection);
}
```

(not sure if this checks for dead connections properly)

Reply:
From what I understand, @marcocitus 's suggested solution makes sense, except that (1) we should probably add the `failed to abort 1PC transaction ...` warning somewhere. (2) Set `transactionState` to `REMOTE_TRANS_ABORTED`.

I tried it on dead connections, and it seems to work.